src/main/java/org/logstash/plugins/outputs/s3/GzipUtil.java

```diff
@@ -0,0 +1,96 @@
+package org.logstash.plugins.outputs.s3;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Ruby Gzip interfaces (such as GzipReader) have limitations to access and manage the ZIP files.
+ * Where Java has a deep control over ZIP files.
+ * This class introduces GZIP related operations.
+ */
+public class GzipUtil {
+
+    private final static int FILE_READ_SIZE = 1024;
+
+    private final static String CORRUPTED_FILE_ERROR = "Unexpected end of ZLIB input stream";
+
+    private final static Logger logger = LogManager.getLogger(GzipUtil.class);
+
+    /**
+     * Compresses a given non-compressed source file into compressed target file.
+     * @param sourcePath a path where source is located
+     * @param targetPath a path where target can be placed
+     * Note: raises an IO exception if any break happens.
+     */
+    public static void compress(String sourcePath, String targetPath) {
+        try {
+            GzipUtil.compress(Paths.get(sourcePath), Paths.get(targetPath));
+        } catch (IOException exception) {
+            logger.error("Error occurred while compressing the file, error={}", exception.getMessage());
+        }
+    }
+
+    /**
+     * Decompresses and recovers corrupted GZIP file into target file.
+     * @param sourcePath a path where corrupted GZIP is located
+     * @param targetPath a path where target can be placed
+     * Note: raises an IO exception other than GZIP dead blocks.
+     */
+    public static void recover(String sourcePath, String targetPath) {
+        try {
+            GzipUtil.decompress(Paths.get(sourcePath), Paths.get(targetPath));
```

Comment:
If the recovery returns `0` bytes, then we should probably skip the rest of the recovery process, to avoid uploading an empty zip file

Reply:
This will covered by [`s.rb:L405~L413`](https://github.com/logstash-plugins/logstash-output-s3/pull/249/files#diff-bdae3d005e86be10100b9ba58d128b55d13c8b8bb4e4ffd43f2c0956c22666e2R405-R412)

Reply:
It is possible for an "empty" gzip file to be created which includes the header but no content, which I don't think will be covered by the above case, and will create an empty recovery file and upload a `gzip` version.

(I could make this happen with `input { generator { count => 250 }}`)

Reply:
Thanks, confirmed/tested possible empty zip cases and updated the PR.